Library/Homebrew/utils/curl.rb

```diff
@@ -198,20 +198,49 @@ def curl_output(*args, **options)
     end
 
     # Check if a URL is protected by CloudFlare (e.g. badlion.net and jaxx.io).
+    # @param details [Hash] Response information from
+    #  `#curl_http_content_headers_and_checksum`.
+    # @return [true, false] Whether a response contains headers indicating that
+    #   the URL is protected by Cloudflare.
+    sig { params(details: T::Hash[Symbol, T.untyped]).returns(T::Boolean) }
     def url_protected_by_cloudflare?(details)
       return false if details[:headers].blank?
-      return unless [403, 503].include?(details[:status].to_i)
+      return false unless [403, 503].include?(details[:status].to_i)
 
-      details[:headers].fetch("set-cookie", nil)&.match?(/^(__cfduid|__cf_bm)=/i) &&
-        details[:headers].fetch("server", nil)&.match?(/^cloudflare/i)
+      set_cookie = details[:headers]["set-cookie"]
+      has_cloudflare_cookie_header = case set_cookie
+      when String
+        set_cookie.match?(/^(__cfduid|__cf_bm)=/i)
+      when Array
+        set_cookie.any? { |cookie| cookie.match?(/^(__cfduid|__cf_bm)=/i) }
+      else
+        false
+      end
```

Comment:
```suggestion
      set_cookie = Array(details[:headers]["set-cookie"])
      has_cloudflare_cookie_header = set_cookie.compact.any? do |cookie|
        cookie.match?(/^(__cfduid|__cf_bm)=/i)
      end
```

Reply:
Oh neat, I've rarely used `Array()` over the years and simply assumed it worked in a straightforward way (i.e., naively return an array containing the argument), so the explanation in the docs is interesting/unexpected. Granted, I'm not entirely surprised that Ruby would handle an array value (not creating a nested array) or a `nil` value (returning an empty array) in a nice way.

Out of curiosity, do we need the `#compact` call if we only expect to encounter an array of strings or an empty array? That is to say, `details[:headers]["set-cookie"]` can be a string, array of strings, or `nil`, and `Array()` will produce an array of strings (for the former two) or an empty array (for the latter).

Reply:
> Out of curiosity, do we need the #compact call if we only expect to encounter an array of strings or an empty array?

Potentially not but seemed like it might be safer than sorry unless we're going to enforce this with Sorbet.

Reply:
What do you think about modifying the `#url_protected_by_*` methods to accept `status` and `headers` arguments instead of passing in the full `details` hash? This would allow us to use Sorbet to enforce the types of these specific values (something we can't do by typing the `details` hash) and we could safely drop the `#compact` calls.

If we're good with this, I'll modify related type signatures (e.g., `#parse_curl_response`, `#parse_curl_output`) to properly define the return type (instead of using `T.untyped`).

Reply:
Related question, what do you think about the idea of defining a `Response` struct and returning that from `#parse_curl_response` instead of a hash? As it stands, we would have to duplicate the relatively complex type signature of a response hash in a few places, so it would make the type signature of every related method more complex:

```
T::Hash[Symbol, T.any(
    String,
    T::Hash[String, T.any(String, T::Array[String])],
  )],
]
```

Using a struct would help to simplify these type signatures (the above would be replaced by `Response`) and if we make changes to the `Response` struct in the future, we wouldn't have to worry about updating every instance of this type signature.

Reply:
> The struct/class idea is more appropriate for a separate PR but there may be value to the idea.

This seems like a neat idea, agreed not for this PR.